libs/utils/src/postgres_backend_async.rs

```diff
@@ -194,6 +200,39 @@ impl PostgresBackend {
         Ok(self)
     }
 
+    /// Returns an AsyncWrite implementation that wraps all the data written
+    /// to it in CopyData messages, and writes them to the connection
+    ///
+    /// The caller is responsible for sending CopyOutResponse and CopyDone messages.
+    pub fn copyout_writer(&mut self) -> CopyDataWriter {
+        CopyDataWriter { pgb: self }
+    }
+
+    /// A polling function that tries to write all the data from 'buf_out' to the
+    /// underlying stream.
+    fn poll_write_buf(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> Poll<Result<(), std::io::Error>> {
+        while self.buf_out.has_remaining() {
+            match Pin::new(&mut self.stream).poll_write(cx, &self.buf_out.chunk()) {
+                Poll::Ready(Ok(bytes_written)) => {
+                    self.buf_out.advance(bytes_written);
+                }
+                Poll::Ready(Err(err)) => return Poll::Ready(Err(err)),
```

Comment:
Note: `1.64` stabilized [ready!](https://doc.rust-lang.org/stable/std/task/macro.ready.html) macro that would simplify this: https://github.com/rust-lang/rust/blob/stable/RELEASES.md#stabilized-apis-2

if https://github.com/neondatabase/neon/issues/3205 lands before this PR, worth a try, because it simplifies noisy unpacks below too.